File: Leetcode756. Pyramid Transition Matrix/Solution.java
class Solution {
    Map<Integer, List<Integer>> map = new HashMap<>();
    int[][] memo;
    public boolean pyramidTransition(String bottom, List<String> allowed) {
        // make a dictionary of the allowed words
        for(String s : allowed) {
            int key = (s.charAt(0)-'A')*7 + (s.charAt(1)-'A');
            map.computeIfAbsent(key, k->new ArrayList<>()).add(s.charAt(2)-'A');
        }
        memo = new int[20][1<<7];
        return solve(bottom);
    }

    public boolean solve(String row) {
        if(row.length() == 1)
            return true;
        return buildNextRow(row, 0, new StringBuilder());
    }

    // helper to build next row character by character
    boolean buildNextRow(String row, int idx, StringBuilder nextRow) {
        if(idx==row.length()-1)
            return solve(nextRow.toString());

        // get the key of the current row of string
        int key = (row.charAt(idx)-'A')*7 + (row.charAt(idx+1)-'A');
        // if the key is not present in the dictionary return false
        if(!map.containsKey(key))
            return false;
        
        for(int val:map.get(key)) {
            nextRow.append((char)(val+'A'));
            if(buildNextRow(row, idx+1, nextRow))
                return true;
            nextRow.deleteCharAt(nextRow.length()-1);
        }
        return false;
    }
}

File: Leetcode756. Pyramid Transition Matrix/Solution.py
from collections import defaultdict
import itertools

class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        # build a map with allowed words
        T=defaultdict(list)
        for triple in allowed:
            T[(triple[0], triple[1])].append(triple[2])

        memo = {}

        # figure out what the next row will look like
        def solve(row):
            if len(row) == 1:
                return True
            if row in memo:
                return memo[row]

            options = []
            for i in range(len(row)-1):
                # for each pair, check the map T
                key = (row[i], row[i+1])
                # if the pair exists add it to our options
                if key in T:
                    options.append(T[key])

                # no block is allowed to sit there
                else:
                    memo[row]=False
                    return False
            
            for next_row_tuple in itertools.product(*options):
                next_row_str = "".join(next_row_tuple)
                if solve(next_row_str):
                    memo[row]=True
                    return True
            memo[row]=False
            return False

        return solve(bottom)
            
